week7/answers.sql

```diff
@@ -1,3 +1,309 @@
+        ON UPDATE CASCADE
+        ON DELETE CASCADE,
+	CONSTRAINT teammemberchar_fk_characters
+        FOREIGN KEY (character_id)
+        REFERENCES characters (character_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+);
+
+CREATE TABLE items (
+	item_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    name VARCHAR(30),
+    armor INT UNSIGNED,
+    damage INT UNSIGNED
+);
+
+CREATE TABLE inventory (
+	inventory_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    character_id INT UNSIGNED,
+    item_id INT UNSIGNED,
+    CONSTRAINT inventorychar_fk_characters
+        FOREIGN KEY (character_id)
+        REFERENCES characters (character_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE,
+	CONSTRAINT inventoryitem_fk_items
+        FOREIGN KEY (item_id)
+        REFERENCES items (item_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+);
+
+CREATE TABLE equipped (
+	equipped_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    character_id INT UNSIGNED,
+    item_id INT UNSIGNED,
+    CONSTRAINT equippedchar_fk_characters
+        FOREIGN KEY (character_id)
+        REFERENCES characters (character_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE,
+	CONSTRAINT equippeditem_fk_items
+        FOREIGN KEY (item_id)
+        REFERENCES items (item_id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+);
+
+
+CREATE OR REPLACE VIEW character_items AS 
+	SELECT
+		c.character_id,
+		c.name AS character_name,	
+		i.name AS item_name,
+		i.armor,
+		i.damage
+	FROM characters c
+		INNER JOIN inventory inv
+			ON c.character_id = inv.character_id
+		INNER JOIN items i
+			ON inv.item_id = i.item_id
+	UNION
+	SELECT 
+		c.character_id,
+		c.name AS character_name,	
+		i.name AS item_name,
+		i.armor,
+		i.damage
+	FROM characters c
+		INNER JOIN equipped eq
+			ON c.character_id = eq.character_id
+		INNER JOIN items i
+			ON eq.item_id = i.item_id
+	ORDER BY character_id ASC, item_name ASC;
+
+
+CREATE OR REPLACE VIEW team_items AS
+	SELECT
+		t.team_id,
+		t.name AS team_name,
+		i.name AS item_name,
+		i.armor,
+		i.damage
+	FROM teams t
+		INNER JOIN team_members tm
+			ON t.team_id = tm.team_id
+		INNER JOIN inventory inv 
+			ON tm.character_id = inv.character_id
+		INNER JOIN items i
+			ON inv.item_id = i.item_id
+UNION 
+	SELECT
+		t.team_id,
+		t.name AS team_name,
+		i.name AS item_name,
+		i.armor,
+		i.damage
+	FROM teams t
+		INNER JOIN team_members tm
+			ON t.team_id = tm.team_id
+		INNER JOIN equipped eq
+			ON tm.character_id = eq.character_id
+		INNER JOIN items i
+			ON eq.item_id = i.item_id
+		ORDER BY team_id, item_name;
+
+
+DELIMITER ;;
+
+CREATE FUNCTION armor_total(char_id INT UNSIGNED)
+RETURNS INT UNSIGNED
+DETERMINISTIC
+BEGIN
+	DECLARE armor_points_equipped INT UNSIGNED;
+    DECLARE armor_points_cs INT UNSIGNED;
+    
+    SELECT SUM(i.armor) INTO armor_points_equipped
+		FROM characters c
+			INNER JOIN character_stats cs
+				ON c.character_id = cs.character_id
```

Comment:
This can be further simplified by SELECTing straight from the `equipped` table and joining `items` (you are not retrieving any new information from the `characters` table that you don't already have available):

```sql
SELECT SUM(i.armor) INTO armor_points_equipped
    FROM equipped eq
        INNER JOIN items i
            ON eq.item_id = i.item_id
    WHERE eq.character_id = char_id;
```